libvast/src/system/disk_monitor.cpp

```diff
@@ -15,6 +15,8 @@
 #include <caf/settings.hpp>
 #include <caf/typed_event_based_actor.hpp>
 
+#include <filesystem>
```

Comment:
> Thank you for trying to improve this code. I'm not sure that a conversion to use the `std::filesystem` API actually achieves that though.
> 
> The filesystem functions must be implemented in terms of syscalls as well, and I'm pretty sure they will both just call `stat` internally. Calling `stat` twice as often as necessary needs a good justification.
> 
> Is your goal to port the code to a non-posix platform?

I don't have a goal of porting in mind, but not making system calls directly would help in that regard. This was just a drive-by refactoring when reading the code as I noticed the other day that all of `vast::path` and `vast::directory` could be replaced with their `std::filesystem` equivalents. Personally (others may not share this same opinion), I find the `std::filesystem` functions easier to read than C-style code with making a struct that gets filled in by a system call. This is more canonical C++ I'd argue.

If performance is a concern here in the disk monitor, I'm fine closing the MR as this change would result in twice as many `stat` system calls as you pointed out I suspect.

Reply:
On the one hand, performance in the "erase" path is probably not a major concern, since this should only be called relatively infrequently.

On the other hand, doing unnecessary work also does not feel  very C++-like.

From how I understand `std::filesystem`, the intended solution here would probably be to use a directory iterator, since struct `directory_entry` already has member functions `file_size` and `last_write_time`, which are allowed to cache the information. Although as far as I can tell only libcxx currently supports the caching.

Reply:
> On the one hand, performance in the "erase" path is probably not a major concern, since this should only be called relatively infrequently.
> 
> On the other hand, doing unnecessary work also does not feel very C++-like.
> 
> From how I understand `std::filesystem`, the intended solution here would probably be to use a directory iterator, since struct `directory_entry` already has member functions `file_size` and `last_write_time`, which are allowed to cache the information. Although as far as I can tell only libcxx currently supports the caching.

Good observation regarding the the `directory_entry` caching of file status -- I noticed that the other day after this PR opened, but did not yet mention it here. 

Note that https://github.com/tenzir/vast/pull/1411 changes the surrounding structure here to use `std::filesystem` with a `directory_iterator` such that we are working with `directory_entry` types in the loop. This would give rise in this PR to take advantage of that like you mentioned.

I'd say it's a quality of implementation concern with whether `libc++` or `libstdc++` caches the `directory_entry` status info for paths.  I'm glad that you brought up the erase performance here in the `disk_monitor` -- thanks for providing some context.

With that being said, @lava @dominiklohmann, do we want to move forward with this PR after https://github.com/tenzir/vast/pull/1411 lands?

Reply:
I'll leave the decision up to @lava, as he wrote most of the disk monitor and should make this decision.

Reply:
I agree that @lava should pull the trigger here.

My 2 cents: assume that the caching takes place and work with `directory_entry`.

Reply:
Yes, I agree, I'm fine with merging this after a rebase.

Reply:
@lava just rebased. Please take a look again. Thanks!